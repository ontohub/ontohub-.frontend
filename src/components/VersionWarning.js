import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Message } from 'semantic-ui-react'
import { Block } from 'jsxstyle'
import semver from 'semver'

const messageStyles = {
  position: 'fixed !important',
  margin: '-1px !important',
  borderRadius: '0 !important',
  bottom: 0,
  width: 'calc(100% + 2px) !important'
}

const WarningMessage = (props) => (
  <Message
    as={Block}
    {...messageStyles}
    negative
    icon="warning sign"
    {...props}
  />
)

export { WarningMessage }

export class VersionWarning extends Component {
  render() {
    if (this.props.error) {
      return (
        <WarningMessage
          header="Could not determine backend version"
          content="This could mean that the backend is currently offline"
        />
      )
    }
    const validVersion =
      this.props.loading ||
      semver.satisfies(this.props.version, this.props.requirement)

    return validVersion
      ? null
      : <WarningMessage
          header="The connected backend does not meet the version requirement"
          content={
            <p>
              Expected version
              {' '}
              <code>{this.props.version}</code>
              {' '}
              to satisfy requirement
              {' '}
              <code>{this.props.requirement}</code>
              . Be aware that this may cause problems.
            </p>
          }
        />
  }
}

VersionWarning.propTypes = {
  version: PropTypes.string,
  loading: PropTypes.bool,
  requirement: PropTypes.string.isRequired
}

export default VersionWarning
