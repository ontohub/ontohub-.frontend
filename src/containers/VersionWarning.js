import { getVersionQuery } from '../apollo/queries'
import { graphql } from 'react-apollo'
import { VersionWarning } from '../components'
import _ from 'lodash'

const VersionWarningWithData = graphql(getVersionQuery, {
  props: (props) => ({
    loading: _.get(props, 'data.loading'),
    error: _.get(props, 'data.error'),
    version: _.at(props, [
      'data.version.tag',
      'data.version.commitsSinceTag'
    ]).join('-'),
    requirement: props.ownProps.requirement
  })
})(VersionWarning)

export default VersionWarningWithData
