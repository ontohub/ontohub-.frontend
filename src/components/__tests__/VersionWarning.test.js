import React from 'react'
import { WarningMessage, VersionWarning } from '../VersionWarning'
import { shallow } from 'enzyme'

describe('WarningMessage', () => {
  it('matches the snapshot', () => {
    const component = shallow(
      <WarningMessage header="Message header" content="Message content" />
    )
    expect(component.getNode()).toMatchSnapshot()
  })
})

describe('VersionWarning', () => {
  let requirement = '> v0.0.0-65'

  describe('version could not be determined', () => {
    it('shows an error message', () => {
      const component = shallow(
        <VersionWarning requirement={requirement} error />
      )
      expect(component.getNode()).toMatchSnapshot()
    })
  })

  describe('version satisfies requirement', () => {
    let version = 'v0.0.0-88'

    it('is hidden', () => {
      const component = shallow(
        <VersionWarning requirement={requirement} version={version} />
      )
      expect(component.getNode()).toBe(null)
    })
  })

  describe('version does not satisfy requirement', () => {
    let version = 'v0.0.0-65'

    it('matches the snapshot', () => {
      const component = shallow(
        <VersionWarning requirement={requirement} version={version} />
      )
      expect(component.getNode()).toMatchSnapshot()
    })
  })
})
