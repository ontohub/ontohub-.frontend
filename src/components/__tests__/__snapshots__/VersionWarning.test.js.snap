// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VersionWarning version could not be determined shows an error message 1`] = `
<WarningMessage
  content="This could mean that the backend is currently offline"
  header="Could not determine backend version"
/>
`;

exports[`VersionWarning version does not satisfy requirement matches the snapshot 1`] = `
<WarningMessage
  content={
    <p>
      Expected version
       
      <code>
        v0.0.0-65
      </code>
       
      to satisfy requirement
       
      <code>
        &gt; v0.0.0-65
      </code>
      . Be aware that this may cause problems.
    </p>
  }
  header="The connected backend does not meet the version requirement"
/>
`;

exports[`WarningMessage matches the snapshot 1`] = `
<Message
  as={[Function]}
  borderRadius="0 !important"
  bottom={0}
  content="Message content"
  header="Message header"
  icon="warning sign"
  margin="-1px !important"
  negative={true}
  position="fixed !important"
  width="calc(100% + 2px) !important"
/>
`;
